components/db/events.go

```diff
@@ -0,0 +1,247 @@
+		tableName,
+		ColumnTimestamp,
+		ColumnDataSource,
+		ColumnEventType,
+		ColumnEventID1,
+		ColumnEventID2,
+		ColumnEventDetails,
+	),
+		ev.Timestamp,
+		ev.DataSource,
+		ev.EventType,
+		ev.EventID1,
+		ev.EventID2,
+		ev.EventDetails,
+	)
+	sqlite.RecordInsertUpdate(time.Since(start).Seconds())
+
+	return err
+}
+
+func findEvent(ctx context.Context, db *sql.DB, tableName string, ev Event) (*Event, error) {
+	selectStatement := fmt.Sprintf(`
+SELECT %s, %s, %s, %s, %s, %s FROM %s WHERE %s = ? AND %s = ? AND %s = ?`,
+		ColumnTimestamp,
+		ColumnDataSource,
+		ColumnEventType,
+		ColumnEventID1,
+		ColumnEventID2,
+		ColumnEventDetails,
+		tableName,
+		ColumnTimestamp,
+		ColumnDataSource,
+		ColumnEventType,
+	)
+	if ev.EventID1 != "" {
+		selectStatement += fmt.Sprintf(" AND %s = ?", ColumnEventID1)
+	}
+	if ev.EventID2 != "" {
+		selectStatement += fmt.Sprintf(" AND %s = ?", ColumnEventID2)
+	}
+
+	params := []any{ev.Timestamp, ev.DataSource, ev.EventType}
+	if ev.EventID1 != "" {
+		params = append(params, ev.EventID1)
+	}
+	if ev.EventID2 != "" {
+		params = append(params, ev.EventID2)
+	}
+
+	var foundEvent Event
+	var eventID1 sql.NullString
+	var eventID2 sql.NullString
+	if err := db.QueryRowContext(ctx, selectStatement, params...).Scan(
+		&foundEvent.Timestamp,
+		&foundEvent.DataSource,
+		&foundEvent.EventType,
+		&eventID1,
+		&eventID2,
+		&foundEvent.EventDetails,
+	); err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	if eventID1.Valid {
+		foundEvent.EventID1 = eventID1.String
+	}
+	if eventID2.Valid {
+		foundEvent.EventID2 = eventID2.String
+	}
+
+	return &foundEvent, nil
+}
+
+// Returns the event in the descending order of timestamp (latest event first).
+func getEvents(ctx context.Context, db *sql.DB, tableName string, since time.Time) ([]Event, error) {
+	query := fmt.Sprintf(`SELECT %s, %s, %s, %s, %s, %s
+FROM %s
+WHERE %s > ?
+ORDER BY %s DESC`,
+		ColumnTimestamp, ColumnDataSource, ColumnEventType, ColumnEventID1, ColumnEventID2, ColumnEventDetails,
+		tableName,
+		ColumnTimestamp,
+		ColumnTimestamp,
+	)
+	params := []any{since.UTC().Unix()}
+
+	start := time.Now()
+	rows, err := db.QueryContext(ctx, query, params...)
+	sqlite.RecordSelect(time.Since(start).Seconds())
+
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return nil, nil
+		}
+		return nil, err
+	}
+	defer rows.Close()
+
+	events := []Event{}
+	for rows.Next() {
+		var event Event
+		var eventID1 sql.NullString
+		var eventID2 sql.NullString
+		if err := rows.Scan(
+			&event.Timestamp,
+			&event.DataSource,
+			&event.EventType,
+			&eventID1,
+			&eventID2,
+			&event.EventDetails,
+		); err != nil {
+			return nil, err
+		}
+		if eventID1.Valid {
+			event.EventID1 = eventID1.String
+		}
+		if eventID2.Valid {
+			event.EventID2 = eventID2.String
+		}
+		events = append(events, event)
```

Comment:
wrap this to a function to avoid duplicate code

Reply:
Addressed in https://github.com/leptonai/gpud/pull/332